cmd/noms/noms_sync.go

```diff
@@ -77,10 +77,21 @@ func runSync(args []string) int {
 		lastProgressCh <- last
 	}()
 
+	sourceRef := types.NewRef(sourceObj)
+	sinkRef, _ := sinkDataset.MaybeHeadRef()
+	nonFF := false
 	err = d.Try(func() {
 		defer profile.MaybeStartProfile().Stop()
+		sinkDataset.Pull(sourceStore, sourceRef, p, progressCh)
+
 		var err error
-		sinkDataset, err = sinkDataset.Pull(sourceStore, types.NewRef(sourceObj), p, progressCh)
+		if sourceRef.Height() > sinkRef.Height() {
```

Comment:
On one hand, it's seemed a bit inconsistent to me that `Dataset.Pull()` handled updating the Head of the dataset and `datas.Pull()` only moved objects around. And I think that we need that flexibility at the `datas` layer. So, for that reason I'd rather make the two Pull methods be more similar.

That said, I think that adding more API to Dataset beyond FastForward() and SetHead() doesn't make much sense. It should have primitives to allow people to implement the policies they want, but not have a helper function for all the different variations of head-changing behavior people might want.

It probably makes sense to add this ref-height check to FastForward(), though. I don't think it makes sense to, for example, have SetHead() always call FastForward() and then follow its own logic if it gets back ErrMergeNeeded.

I think the problem is that we still don't have a great way of testing logic in our command line tools, which is something we need to fix